File: Semestr3/Paradygmaty-Programowania/Cwiczenia/Cwiczenia3/zadania3.scala

object zadania3 extends App {
  // zad 2
  def curry3[A, B, C, D](f: ((A, B, C)) => D): A => B => C => D =
    (a: A) => (b: B) => (c: C) => f((a, b, c))

  val curry3 = [A, B, C, D] =>
    (f: ((A, B, C)) => D) => (a: A) => (b: B) => (c: C) => f((a, b, c))

  def uncurry3[A, B, C, D](f: A => B => C => D): ((A, B, C)) => D = {
    case (a, b, c) => f(a)(b)(c)
  }

  val uncurry3 = [A, B, C, D] =>
    (f: A => B => C => D) => (t: (A, B, C)) => f(t._1)(t._2)(t._3)

//zad 3
  def sumProd(xs: List[Int]): (Int, Int) = {
    xs.foldLeft((0, 1))((accum, elem) =>
      accum.match { case (a, b) => (a + elem, b * elem) }
    )
  }

  val s = sumProd(List(1, 2, 3, 4, 5))

// INSERTION SORT :( MERGE SORT KURWAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
// 1 2 1 4 9 3 ->

  def insertionSort[A](comp: (A, A) => Boolean, xs: List[A]): List[A] = {
    def insert(xs: List[A], elem: A): List[A] = xs match 
        case head :: tail =>
          if (!comp(elem, head)) elem :: xs
          else head :: insert(tail, elem)
        case Nil => List(elem)
      

    def sort(xs: List[A], sortedPart: List[A]): List[A] = xs match
      case head :: tail => sort(tail, insert(sortedPart, head))
      case Nil          => sortedPart

    sort(xs, List())
  }


  def mergeSort[A] (comp: (A, A) => Boolean, xs: List[A]): List[A] = {
    def merge (comp: (A, A) => Boolean, result: List[A], xs: List[A], ys: List[A]): List[A] = {
        (xs, ys) match
            case (h1::t1, h2::t2) =>
                 if (comp(h1, h2)|| !comp(h1,h2)) then merge(comp, h1::result, t1, ys)
                 else  merge(comp, h2::result, xs, t2)
            case ( Nil, h2::t2 ) => result.reverse ::: ys
            case ( h1::t1, Nil ) => result.reverse ::: xs
            case (Nil, Nil) => result.reverse
    }
    def partition (xs: List[A]): (List[A], List[A]) = {
        def iter (xs: List[A], fast : List[A], slow : List[A], left: List[A], right: List[A]): (List[A], List[A]) = {
            (fast, slow) match
                case (h1::a::b, h2::t2) => iter(xs, b, t2, h2::left,h2:: t2)
                case (h1::Nil, h2::t2 ) => ((left).reverse,h2:: t2)
                case (Nil, h2::t2) => ((left).reverse, h2::t2)
                case _ =>  ((left).reverse, Nil)
            } 
        iter(xs, xs , xs, Nil, Nil)
    }
    def sort (xs : List[A]) : List[A] = {
        xs match
            case Nil => Nil
            case List(_) => xs
            case _ => {
                val (a,b) = partition(xs)
                merge (comp, Nil , sort(a) , sort(b))
            }
        }
        sort(xs)
    }

   // val sortTest = mergeSort((a:Int, b:Int) => a>=b, List(3,2,3,9,1) )
    //print (sortTest.toString())


  println("1️⃣ Prosty rosnący sort:")
  println(mergeSort((a: Int, b: Int) => a < b, List(3, 1, 2)))
  println("Oczekiwane: List(1, 2, 3)\n")

  println("2️⃣ Nieparzysta liczba elementów:")
  println(mergeSort((a: Int, b: Int) => a < b, List(5, 4, 3, 2, 1)))
  println("Oczekiwane: List(1, 2, 3, 4, 5)\n")

  println("3️⃣ Parzysta liczba elementów:")
  println(mergeSort((a: Int, b: Int) => a < b, List(6, 5, 4, 3, 2, 1)))
  println("Oczekiwane: List(1, 2, 3, 4, 5, 6)\n")

  println("4️⃣ Duplikaty:")
  println(mergeSort((a: Int, b: Int) => a < b, List(3, 1, 3, 2, 3)))
  println("Oczekiwane: List(1, 2, 3, 3, 3)\n")

  println("5️⃣ Już posortowana:")
  println(mergeSort((a: Int, b: Int) => a < b, List(1, 2, 3, 4, 5)))
  println("Oczekiwane: List(1, 2, 3, 4, 5)\n")

  println("6️⃣ Odwrotne sortowanie:")
  println(mergeSort((a: Int, b: Int) => a > b, List(1, 2, 3, 4, 5)))
  println("Oczekiwane: List(5, 4, 3, 2, 1)\n")

  println("7️⃣ Dłuższa lista losowa:")
  println(mergeSort((a: Int, b: Int) => a < b, List(10, 3, 5, 1, 9, 4, 2, 8, 6, 7)))
  println("Oczekiwane: List(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)\n")

  println("8️⃣ Jednoelementowa lista:")
  println(mergeSort((a: Int, b: Int) => a < b, List(42)))
  println("Oczekiwane: List(42)\n")

  println("9️⃣ Pusta lista:")
  println(mergeSort((a: Int, b: Int) => a < b, Nil))
  println("Oczekiwane: List()\n")

  println("🔟 Ujemne i dodatnie liczby:")
  println(mergeSort((a: Int, b: Int) => a < b, List(0, -5, 3, -2, 1, -4)))
  println("Oczekiwane: List(-5, -4, -2, 0, 1, 3)\n")
}

File: Semestr3/Paradygmaty-Programowania/Cwiczenia/Cwiczenia3/Zadania3.ml
(* Podaj (i uzasadnij!) najogólniejsze typy poniższych funkcji (samodzielnie, bez pomocy kompilatora
OCamla!) :
 a) let f1 x = x 2 2;; b) let f2 x y z = x ( y ^ z );; *)

 (*
 A) funkcja x przyjamuje argument int i zwraca funkcje ktora przyjmuje także argument int i zwraca "coś" czyli:
 x: int ->( int -> a ) = int -> int -> a
 typ: (int -> int -> a) -> a bo f1 zwraca to samo co x

 B) operator ^ łączy stringi czyli y -> string i z -> string czyli argumentem funkcji x jest string czyli:
 x: string -> a
 funckja f2 jest trójargumentowa i zwraca taki sam typ jak x
 (string -> a) -> string -> string -> a
 *)

 (*
 2. Zdefiniuj funkcje a) curry3 i b) uncurry3, przeprowadzające konwersję między zwiniętymi i rozwiniętymi
 postaciami funkcji od trzech argumentów. Podaj (i uzasadnij!) ich typy.
 Dla każdej funkcji napisz dwie wersje: z lukrem syntaktycznym i bez lukru syntaktycznego
 *)


 let curry3 f a b c = f (a,b,c);;
 let curry3 = fun f -> fun a -> fun b -> fun c -> f (a,b,c);;
 let uncurry3 f(a,b,c) = f a b c;;
 let uncurry3 = fun f -> fun(a,b,c)  -> f a b c

 (*Sprawdzian 2 na eportalu wrzucic*)
 let curry4 f a b c d = f(a,b,c,d) ;; (* lukier*)
 let curry4 = fun f-> fun a -> fun b -> fun c -> fun d -> f(a,b,c,d);; (*brak lukru*)
 let uncurry4 f(a,b,c,d) = f a b c d;; (*lukier*)
 let uncurry4 = fun f -> fun(a,b,c, d)  -> f a b c d;; (*brak lukru*)


 (*
 przekształć poniższą rekurencyjną definicję funkcji sumProd, która oblicza jednocześnie
 sumę i iloczyn listy liczb całkowitych na równoważną definicję nierekurencyjną z
 jednokrotnym użyciem funkcji bibliotecznej fold_left (Scala – foldLeft), której argumentem jest
 odpowiednia funkcja anonimowa (literał funkcyjny).
 *)

 let rec sumProd xs = 
   match xs with 
   | h::t -> let (s,p) = sumProd t in (h+s, h*p)
   | [] -> (0,1);;
let sumProd xs = List.fold_left (fun accum elem -> match accum with (a,b) -> (a+ elem, elem * b)) (0,1) xs;;

(*zadanie 4*)

let rec quicksort = function
 [] -> []
 | [x] -> [x]
 | xs -> let small = List.filter (fun y -> y < List.hd xs ) xs
 and large = List.filter (fun y -> y >= List.hd xs ) xs
 in quicksort small @ quicksort large;;

 (* W niektórych przypadkach poniższy algorytm nie działa poprawnie, ponieważ filtruje listy
 wraz z pivotem. W efekcie pivot z poprzedniego wywołania może trafić do listy `large`. 

 W poprawnym wariancie algorytmu pivot powinien być jedynie użyty do porównywania elementów
 (czyli do filtrowania), ale nie powinien znajdować się już w filtrowanej liście.

 Następnie pivot należy umieścić pomiędzy posortowaną listą `small` i posortowaną listą `large`
 w wyniku końcowym. *)
let s = quicksort [5;4;3];;

let rec quicksort' = function
 [] -> []
 | x::xs -> let small = List.filter (fun y -> y < x ) xs
 and large = List.filter (fun y -> y > x ) xs
 in quicksort' small @ (x :: quicksort' large);;

 let ss = quicksort'[6;6;9];;

 (*Ta wersja nie działa poprawnie w przypadku list zawierających powtarzające się wartości.
 Lista `large` jest tworzona przy użyciu filtra `(fun y -> y > x)`, 
 przez co elementy równe pivotowi `x` nie trafiają ani do `small`, ani do `large`
 i w efekcie zostają zgubione.*)


let insertionSort comp xs =
  let rec insert xs elem =
    match xs with
    | head :: tail ->
        if comp elem head then elem :: xs
        else head :: insert tail elem
    | [] -> [elem]
  in
  let rec sort xs sortedPart =
    match xs with
    | head :: tail -> sort tail (insert sortedPart head)
    | [] -> sortedPart
  in
  sort xs []
;;

let comp (key1, _) (key2, _) = key1 < key2;;
let stInser = insertionSort comp [(2,'C');(1,'A'); (1,'B')];;

let mergeSort comp xs =
  let rec merge comp result xs ys =
    match (xs, ys) with
    | (h1 :: t1, h2 :: t2) ->
        if comp h1 h2 || not (comp h2 h1) then
          merge comp (h1 :: result) t1 ys
        else
          merge comp (h2 :: result) xs t2
    | ([], h2 :: t2) -> List.rev result @ ys
    | (h1 :: t1, []) -> List.rev result @ xs
    | ([], []) -> List.rev result
  in
  let partition xs =
    let rec iter xs fast slow left right =
      match (fast, slow) with
      | (h1 :: a :: b, h2 :: t2) -> iter xs b t2 (h2 :: left) (h2 :: t2)
      | (h1 :: [], h2 :: t2) -> (List.rev left, h2 :: t2)
      | ([], h2 :: t2) -> (List.rev left, h2 :: t2)
      | _ -> (List.rev left, [])
    in
    iter xs xs xs [] []
  in
  let rec sort xs =
    match xs with
    | [] -> []
    | [_] -> xs
    | _ ->
        let (a, b) = partition xs in
        merge comp [] (sort a) (sort b)
  in
  sort xs
;;

let s = mergeSort (fun a b -> a<b ) [2;4;6;3;1;2;1];;
let s1 = mergeSort (fun a b -> a<b ) [2;4];;
let s2 = mergeSort (fun a b -> a<b ) [2;4;6];;
let s3 = mergeSort (fun a b -> a<b ) [2;4;6;3;1;2;1];;
let st = mergeSort comp [(2,'C');(1,'A'); (1,'B')];;
      



